### DISCUSSION TITLE
Scalability/performance testing plan

### DISCUSSION THREAD

**Discussion by dejanb** [guacsec/trustify#214] [Security]
We started with basic scalability testing with manually ingesting pre-planned number of documents and then again manually verifying that UI is responsive. This is great effort. Going forward it would be great to take this practice to the next level and being able to automate this kind of testing and get exact results that we can compare with previous tests, versions and deployment options. This discussion is to collect ideas of what we wanna do, how and share currently known best practices.

**Comment by dejanb** [comment]
I think there are a few tasks that we can perform right now, like:

* document current procedure for data ingestion
* Automate data ingestion, preferably dockerized workload that can be run locally or deployed as a job in the cluster
* Start working on API/UI testing client that we can use for testing

Open questions:
* How to create client API? Do we already have what we need from other testing methods?
* What are we testing? Throughput? Response time? Resources?
* Do we need any kind of tool for this, like JMeter? Are there any better "cloud-native" tools available?

**Reply by helio-frota** [reply]
> What are we testing? Throughput? Response time? Resources?

> Are there any better "cloud-native" tools available?

yes

for api endpoints we can [probably] eat own dog food with hyperfoil 
https://github.com/Hyperfoil


https://github.com/obs-nebula/hyperfoil-test

**Reply by helio-frota** [reply]
![2024-04-29_14-10](https://github.com/trustification/trustify/assets/6443576/91caa6af-65e9-4f6b-bbd4-5f0367e33447)

**Comment by gildub** [comment]
Regarding data set loading, from the volume testing experience, using cronjobs ([follow the same approach as post-installation](https://github.com/trustification/trustification/blob/main/docs/modules/admin/pages/cluster-post-install.adoc)) is likely the most efficient and consistent way of loading data in RHTPA and it can be entirely automated. For each category of data set (i.e CSAF, SBOM, CEV, etc) the test environment must provide the data set to be available as a URL.

The other alternative which is a combination of a host (pod or remote machine) running the file walkers with data set sources (files or URL) are either less efficient or more complex to automate.

No matter the chosen solution we need to have a place where we maintain consistent CSAF, SBOM, CEV, Packages, data sets which will be used those automated tests.

**Comment by helio-frota** [comment]
We have UI and soon sample data for PM-mode https://github.com/trustification/trustify/pull/297
`UI` + `sample data` + hyperfoil (`in-vm`) sounds like a good thing to create some **scenarios** in case team wants to follow this path. (probably)

**Reply by helio-frota** [reply]
or 
UI + sample data +  `Drill` Like Jim mentioned above

**Reply by helio-frota** [reply]
I found something with hyperfoil and ipv6

![2024-05-15_15-00](https://github.com/trustification/trustify/assets/6443576/25f9fe1f-a0b7-433b-a5ca-812b13fd9d72)


After changing the code [here](https://github.com/trustification/trustify/blob/main/common/infrastructure/src/app/http.rs#L547)
to this: `None => http.bind(("127.0.0.1", 8080)).context("bind")?,`

it worked:

![2024-05-15_15-01](https://github.com/trustification/trustify/assets/6443576/95ffe2be-fe2d-4f47-814b-6c4d6bfe8e36)

**Reply by helio-frota** [reply]
@ctron hi, any possibility to use 0.0.0.0:8080 instead of [::1]:8080 ?

![2024-05-15_17-18](https://github.com/trustification/trustify/assets/6443576/7f2741ac-b822-430f-aa83-2e5b6988f16a)

**Reply by helio-frota** [reply]
:thinking: 

`any possibility to use 0.0.0.0:8080 instead of [::1]:8080 ?`

maybe we can't do it ... hyperfoil worked with `trustification` because `localhost:8080` wasn't handled by the same code as the project uses a different architecture.

And we hit [spog](https://github.com/trustification/trustification/blob/99e7f442cff1168cae49dad797e86a1a06022d57/infrastructure/src/endpoint.rs#L129
) first with hyperfoil tests that constructs a [SocketAddr](https://github.com/trustification/trustification/blob/99e7f442cff1168cae49dad797e86a1a06022d57/infrastructure/src/endpoint.rs#L42) with ipv4 format

**Reply by helio-frota** [reply]
So https://github.com/franz1981 found the issue and this works with ipv6 as well

```
[hyperfoil]$ start-local
Starting controller in default directory (/tmp/hyperfoil)
Controller started, listening on 127.0.0.1:45285
Connecting to the controller...
Connected to 127.0.0.1:45285!
[hyperfoil@in-vm]$ upload ../hf.yml
Loaded benchmark foo-foil, uploading...
... done.
[hyperfoil@in-vm]$ run
Started run 0001
Run 0001, benchmark foo-foil
Agents: in-vm[STOPPED]
Started: 2024/05/16 13:44:20.591    Terminated: 2024/05/16 13:44:30.601
NAME  STATUS      STARTED       REMAINING  COMPLETED     TOTAL DURATION               DESCRIPTION
main  TERMINATED  13:44:20.592             13:44:30.600  10008 ms (exceeded by 8 ms)  50.00 users per second
[hyperfoil@in-vm]$
```

![2024-05-16_13-52](https://github.com/trustification/trustify/assets/6443576/60f5cc48-8877-40f9-b1c8-882b0eac36a8)

**Reply by franz1981** [reply]
The problem seems related to a system property we set in hyperfoil to forcibly use the ipv4 stack, likely to not break Infinispan clustering.
I have to double check with the data grid team if the problem is legit or I should find a way to fix it differently

**Reply by helio-frota** [reply]
@franz1981 we have a PR ( sort of draft ) https://github.com/trustification/trustify/pull/309 thanks for the support!

**Comment by helio-frota** [comment]
@dejanb Do we have any expectations regarding the number of acceptable sboms or requests?
thanks

**Reply by helio-frota** [reply]
1GiB sboms ?
20k requests ?

**Reply by helio-frota** [reply]
We have this currently:

![2024-05-29_12-50](https://github.com/trustification/trustify/assets/6443576/f9459ba7-a825-4c0a-8485-eeb290e4f84c)

**Reply by dejanb** [reply]
That's a great starting point. I think for starters we should not focus on the huge number of requests, but gather some numbers on response times even for small number of requests.

**Reply by helio-frota** [reply]
Agreed

**Reply by helio-frota** [reply]
@gildub ^

**Comment by helio-frota** [comment]
I think we can close this :+1: 

https://github.com/trustification/trustify-load-test-runs
https://github.com/trustification/scale-testing

Can we close this?
